### PR TITLE
Added accounting for terminal size

### LINKED ISSUE #64: [bug] issues when rendering a select whose items are longer than the active terminal viewport
Just ran into this for the first time -- seems that if the select is long, your cursor will be placed at the bottom while the first selected item is at the top, outside of the active terminal. If you scroll up to see the active item and then press "down" to the next item, it will indeed select the correct item, but the entire list will be re-rendered and the cursor placed again at the bottom. Once the active item comes into the viewport, then it navigates correctly; the issue seems to be when the active item is outside of the viewport.

Hard to explain, easiest reproduced by creating a select with enough items to overflow the active terminal viewport and pressing buttons :sweat_smile: 

Here's a gif:
cliclack_long_select

One thing that would be nice here is to be able to specify a maximum size for the select, i.e. max height 5 lines and then scrolling the items when navigating up/down. In my example above, I don't really want my user's terminal to be spammed with all possible bitcoin core versions, it's most likely one of the most recent versions they're interested in, but I don't want to deny them older versions either. So only displaying i.e. the 5 most recent versions at first would be a benefit.

Not quite sure how to implement this if I were to take a stab at it...

### PR DESCRIPTION
**Pull request by BjornTheProgrammer**
Fixes #64 and #79, and is based upon work in #73, but just refined.

Adds one external API called `set_size` to modify the number of rows displayed at once for `select` and `multiselect`.

By default, this ensures that there are not more than the maximum possible rows displayed at once.

### CODE REVIEW

**Comment by fadeevab** [comment]
@BjornTheProgrammer Thanks! I'll check it out later. As of now, apply `rustfmt` over your changes (Alt+Shift+F in vscode), I can see TABs and some formatting inconsistencies.

**Comment by BjornTheProgrammer** [comment]
Whoops, should be fixed now. I ran `cargo fmt`, which should be the same as VSCode's.

**Comment by BjornTheProgrammer** [comment]
I realized that the test used the wrong function name, I've updated that.

**Comment by BjornTheProgrammer** [comment]
Makes perfect sense! I'll implement the change as requested. Probably should have done this to begin with 😅️.

**Comment by BjornTheProgrammer** [comment]
Should be implemented now, please let me know if you have any other issues!

**Comment by furbyhaxx** [comment]
termsize does not work for SSH terminals, like RustRover trough Jetbrains Gateway.
I made a fix will submit a PR for termsize which fixes this, but until then we could directly integrate a workaround for this case:

```rust
extern crate libc;

use std::io::IsTerminal;

use std::ffi::{c_ushort, CString};

use self::{
    super::Size,
    libc::{ioctl, O_RDONLY, STDOUT_FILENO, TIOCGWINSZ},
};

/// A representation of the size of the current terminal
#[repr(C)]
#[derive(Debug)]
pub struct UnixSize {
    /// number of rows
    pub rows: c_ushort,
    /// number of columns
    pub cols: c_ushort,
    x: c_ushort,
    y: c_ushort,
}


/// Workaround for SSH terminal size
pub fn get_term_size() -> Option<termsize::Size> {
    if cfg!(target_os = "windows") {
        termsize::get()
    } else {
        _get_unix_termsize()
    }
}

/// Gets the current terminal size
fn _get_unix_termsize() -> Option<termsize::Size> {
    // http://rosettacode.org/wiki/Terminal_control/Dimensions#Library:_BSD_libc
    if !std::io::stdout().is_terminal() {
        return None;
    }
    let mut us = UnixSize {
        rows: 0,
        cols: 0,
        x: 0,
        y: 0,
    };

    let fd = if let Ok(ssh_term) = std::env::var("SSH_TTY") {
        // Convert path to a C-compatible string
        let c_path = CString::new(ssh_term).expect("Failed to convert path to CString");

        // Open the terminal device
        let fd = unsafe { libc::open(c_path.as_ptr(), O_RDONLY) };
        if fd < 0 {
            return None; // Failed to open the terminal device
        }

        fd
    } else {
        STDOUT_FILENO
    };

    let r = unsafe { ioctl(fd, TIOCGWINSZ, &mut us) };

    if r == 0 {
        Some(termsize::Size {
            rows: us.rows,
            cols: us.cols,
        })
    } else {
        None
    }
}
```

Here we can just call get_term_size instead of termsize::get() but it introduces a new dependency (libc).

**Comment by BjornTheProgrammer** [comment]
I'm more than willing to add libc as a temporary fix, but it really is  up to whether @fadeevab wants it. If you want to use this patch now, without waiting for your termsize patch to be merged, you could probably override both dependencies in the cargo.toml

**Comment by fadeevab** [comment]
Hey, folk! You're still here 😄
If you want to bring a `libc` dependency, do it in a cross-platform way: the code should compile on Windows.

Other than that, I think the "size" in the PR always actually means "height", or "rows", or "max_height". Like, "term_max_height" or "frame_height"/"frame_max_height".

Overall, I think the "max height" should always be detected automatically. If you want to expose the `set_height` (`set_max_height`?) method, you must test corner cases: What if I resize the terminal to the height of two rows?

**Comment by BjornTheProgrammer** [comment]
>  If you want to bring a libc dependency, do it in a cross-platform way: the code should compile on Windows.

Yup, it could be a target libc with something like
```rust
[target.'cfg(unix)'.dependencies]
libc = "0.2"
```

> Other than that, I think the "size" in the PR always actually means "height", or "rows", or "max_height". Like, "term_max_height" or "frame_height"/"frame_max_height".

Totally agree here, I think the best name would be rows though. Since (at least in unix) that is how it is treated.

> Overall, I think the "max height" should always be detected automatically. If you want to expose the set_height (set_max_height?) method, you must test corner cases: What if I resize the terminal to the height of two rows?

I'll do some more testing, but thanks to @furbyhaxx the smaller than 3 corner cases should be handled.

> Overall, I think the "max height" should always be detected automatically. If you want to expose the set_height (set_max_height?) method, you must test corner cases: What if I resize the terminal to the height of two rows?

I believe this is already the case? The default function detects the number of rows, and sets the max height to that by default.

**Comment by furbyhaxx** [comment]
We should also verify that your -3 margin is always a sane default.

I needed to change it to 5 to get a select with filter completely fit the terminal. But could be because of the type of terminal I use currently.

**Comment by BjornTheProgrammer** [comment]
Yeah, the 3 was there since that is just for the "Select an Item", the search, and the final empty closing line. Probably should at the very least include one or two items.

Other than that I just made a commit, which compiles for unix and windows, which should theoretically fix the ssh issue. Mind giving it a try? Anyone who has a windows computer should probably test the new commit as well.

**Comment by fadeevab** [comment]
Hi guys, a busy week, I remember about the PR, I'll try to get back to it "ASAP"

File: src/prompt/term.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+impl Default for TermSize {
+    fn default() -> Self {
+        let mut window_max_rows = usize::MAX;
+
+        if let Some(termsize) = get_term_size() {
+            window_max_rows = (termsize.rows as usize)
+                .checked_sub(3)
+                .unwrap_or(termsize.rows as usize);
+        }
```
Comment (fadeevab) [inline src/prompt/term.rs:14]: I realized that it just not needed. Try to shrink your terminal to a few rows of height, and then run the `cargo run --example filter`:  the list is gonna be forever of a few rows, even if you expand the terminal height.

So, with `max_rows = usize::MAX` by default, the lists are gonna be working in an old way, which is good for backward compatibility.

**Review by fadeevab — CHANGES_REQUESTED** [review]
@BjornTheProgrammer We're gonna simplify and minimize your patch! :) See review comments. Thank you for your contribution!

**Comment by fadeevab** [comment]
Closed in favor of PR #91

**Comment by fadeevab** [comment]
Selection lists now have a scrolling behavior after setting a `max_rows` number. Thanks @BjornTheProgrammer for providing this PR, which is mainly used for the implementation, I just polished it and refactored.
